cmd/core/node_association.go

```diff
@@ -1,15 +1,67 @@
 package core
 
-type SessionMap map[uint64]Session
+import (
+	"context"
+	"github.com/edgecomllc/eupf/cmd/config"
+	"time"
+)
 
 type NodeAssociation struct {
-	ID            string
-	Addr          string
-	NextSessionID uint64
-	Sessions      SessionMap
+	ID               string
+	Addr             string
+	NextSessionID    uint64
+	Sessions         map[uint64]*Session
+	HeartbeatRetries uint32
+	cancelRetries    context.CancelFunc
+}
+
+func NewNodeAssociation(remoteNodeID string, addr string) *NodeAssociation {
+	return &NodeAssociation{
+		ID:            remoteNodeID,
+		Addr:          addr,
+		NextSessionID: 1,
+		Sessions:      make(map[uint64]*Session),
+	}
 }
 
 func (association *NodeAssociation) NewLocalSEID() uint64 {
 	association.NextSessionID += 1
 	return association.NextSessionID
 }
+
+func (association *NodeAssociation) RefreshRetries() {
+	association.HeartbeatRetries = 0
+	if association.cancelRetries != nil {
+		association.cancelRetries()
+	}
+	association.cancelRetries = nil
+}
+
+func (association *NodeAssociation) IsExpired() bool {
+	return association.HeartbeatRetries > config.Conf.HeartbeatRetries
+}
+
+func (association *NodeAssociation) IsHeartbeatScheduled() bool {
+	return association.cancelRetries != nil
+}
+
+// ScheduleHeartbeatRequest schedules a series of heartbeat requests to be sent to the remote node. Return a cancellation function to stop the scheduled requests.
+func ScheduleHeartbeatRequest(duration time.Duration, conn *PfcpConnection, association string) context.CancelFunc {
```

Comment:
Seems like `ScheduleHeartbeatRequest` and `SendHeartbeatRequest` belong to accosiation type and have to be methods of this type